libpd_wrapper/util/ringbuffer.h

```diff
@@ -8,42 +8,60 @@
  *
  */
 
-#ifndef __OPENSL_RING_BUFFER_H__
-#define __OPENSL_RING_BUFFER_H__
+#ifndef __RING_BUFFER_H__
+#define __RING_BUFFER_H__
 
-// Simple lock-free ring buffer implementation for one writer thread and one
-// consumer thread.
+/// Simple lock-free ring buffer implementation for one writer thread and one
+/// consumer thread.
 typedef struct ring_buffer {
     int size;
     char *buf_ptr;
     int write_idx;
     int read_idx;
+    int atomic;
 } ring_buffer;
 
-// Creates a ring buffer (returns NULL on failure).
+/// Creates a ring buffer (returns NULL on failure).
+/// Size must be multiple of 256.
 ring_buffer *rb_create(int size);
 
-// Deletes a ring buffer.
+/// Deletes a ring buffer.
 void rb_free(ring_buffer *buffer);
 
-// Returns the number of bytes that can currently be written; safe to be called
-// from any thread.
+/// Returns the number of bytes that can currently be written; safe to be called
+/// from any thread.
 int rb_available_to_write(ring_buffer *buffer);
 
-// Returns the number of bytes that can currently be read; safe to be called
-// from any thread.
+/// Returns the number of bytes that can currently be read; safe to be called
+/// from any thread.
 int rb_available_to_read(ring_buffer *buffer);
 
-// Writes bytes from n sources to the ring buffer (if the ring buffer has
-// enough space). The varargs are pairs of type (const char*, int) giving a
-// pointer to a buffer and the number of bytes to be copied. Only to be called
-// from a single writer thread.
-// Returns 0 on success.
+/// Writes bytes from n sources to the ring buffer (if the ring buffer has
+/// enough space). The varargs are pairs of type (const char*, int) giving a
+/// pointer to a buffer and the number of bytes to be copied. Only to be called
+/// from a single writer thread.
+/// Returns 0 on success.
 int rb_write_to_buffer(ring_buffer *buffer, int n, ...);
 
-// Reads the given number of bytes fromthe ring buffer to dest if the ring
-// buffer has enough data. Only to be called from a single reader thread.
-// Returns 0 on success.
+/// Writes single byte value n times to the ring bufer (if the ring buffer has
+/// enough space.) Only to be called from a single writer thread.
+/// Returns 0 on success.
+int rb_write_value_to_buffer(ring_buffer *buffer, int value, int n);
+
+/// Reads the given number of bytes from the ring buffer to dest if the ring
+/// buffer has enough data. Only to be called from a single reader thread.
+/// Returns 0 on success.
 int rb_read_from_buffer(ring_buffer *buffer, char *dest, int len);
 
+/// Clears the contents of the ring buffer; safe to be called from any thread.
+void rb_clear_buffer(ring_buffer *buffer);
+
+/// Set the atomicity of the buffer. By default, the buffer is atomic and is
+/// inherently thread safe. If you are only using one thread to read & write,
+/// disabling this can increase performance slightly.
```

Comment:
I generally try to avoid setters like this because they're error prone, in terms of implementation as well as usage.

Did you measure the performance, with and without this optimization? I would expect that this won't make a significant difference.

Reply:
Maybe. It just made sense to me as I didn't need the locking within the AudioUnit. I figured it could be just an unsafe set as it's not something you change often. I have not checked performance and maybe it doesn't make much of a difference but there is an added comparison for the atomicity now.

Here's the usage: https://github.com/libpd/libpd/blob/update/buffered-audiounit2/objc/PdBufferedAudioUnit.m

Reply:
On second thought, how about if the feature was kept but was only set upon creation ala a second creator function?

    ring_buffer *rb_create(int size, int atomic);

Reply:
In regards to performance, I've not made an explicit test, but just from the convention of nonatomic over atomic Obj-C property attributes and a little poking around: http://demin.ws/blog/english/2012/05/05/atomic-spinlock-mutex/

My reasoning was: Why need to sync/fetch if you don't need to at the cost of 1 comparison? It didn't take much work to add it.